### PR TITLE
Issue #1

### PR DESCRIPTION
**Pull request by gma1083**
Added validate method to timesheets class with accompanying tests.

### CODE REVIEW

File: source/timesheet.js [hunk 1]
```diff
@@ -19,23 +21,30 @@ class Timesheet{
         this.notes = timesheet.notes;
     }
 
-    save() {
+    async save() {
```
Comment (cucucachu) [inline source/timesheet.js:24]: Adding the async keyword here is necessary but unrelated to this issue.

File: source/timesheet.js [hunk 2]
```diff
@@ -19,23 +21,30 @@ class Timesheet{
...
         return db.insertOne(this, collectionName);
     }
 
-    delete() {
+    async delete() {
```
Comment (cucucachu) [inline source/timesheet.js:28]: Likekwise this is unrelated to this issue.

File: source/timesheet.js [hunk 3]
```diff
@@ -19,23 +21,30 @@ class Timesheet{
...
+        await db.updateOne({_id : this.jobID}, { $pull: { timsheets : this.jobID} }, 'Jobs');
```
Comment (cucucachu) [inline source/timesheet.js:29]: Likekwise this is unrelated to this issue.

File: source/timesheet.js [hunk 4]
```diff
@@ -19,23 +21,30 @@ class Timesheet{
...
-    validate() {
-        if(!(this._id instanceof mongodb.ObjectID)) throw new Error('timeSheetID is not valid');
-        if(!(this.date instanceof Date)) throw new Error('Date is not valid');
-        if(!(this.employeeID instanceof mongodb.ObjectID)) throw new Error('EmployeeID is not valid');
-        // if(typeof(this.employee) !== 'string') throw new Error('Street Name is not valid');
-        // if(typeof(this.city) !== 'string') throw new Error('City is not valid');
-        // if(typeof(this.state) !== 'string') throw new Error('State is not valid');
-        // if(typeof(this.zip) !== 'number') throw new Error('Zip is not valid');
-        // if(typeof(this.county) !== 'string') throw new Error('Country is not valid');
+    async validate() {
```
Comment (cucucachu) [inline source/timesheet.js]: validate should not be asynchronous because you're not calling any asynchronous code.

File: source/timesheet.js [hunk 5]
```diff
@@ -19,23 +21,30 @@ class Timesheet{
...
-    save() {
+    async save() {
         return db.insertOne(this, collectionName);
     }
 
-    delete() {
+    async delete() {
+        await db.updateOne({_id : this.jobID}, { $pull: { timsheets : this.jobID} }, 'Jobs');
         return db.deleteOne({_id : this._id}, collectionName);
     }
```
Comment (cucucachu) [inline source/timesheet.js:32]: Please add a method comment for this method.
Should be in the form

Method Name
Parameters
Returns
Throws

File: test/job.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ describe('Job.js Tests:', function(){
     });
 
 
-    describe('Job Constructor Tests:', async function(){
+    describe('Job Constructor Tests:', function(){
```
Comment (cucucachu) [inline test/job.js]: All the changes on this file are not related to this issue. 

Please revert the file back to its original state in this branch.

File: test/timesheet.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+            tmCrewSize : 0,
+            tmCrewHours : 0,
+            lunchTaken : true,
+            jobFinished : true,
+            offHauled : false,
+            yardsHauled : 0,
+            notes : 'Test Notes'
+        };
+
+        it('Timesheet Validation - Happy Path', async function(){
```
Comment (cucucachu) [inline test/timesheet.js]: All of these 'it' s do not need to be asynchronous once you change validate() to not be asynchronous.

File: test/timesheet.js [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+            let timesheet = new Timesheet(timesheetValues);
+            await timesheet.validate();
```
Comment (cucucachu) [inline test/timesheet.js]: Likewise, you will not need to 'await' validate() once it is no longer asynchronous.

File: test/timesheet.js [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+        });
+
+        it('Timesheet Validation - _id', async function(){
```
Comment (cucucachu) [inline test/timesheet.js]: All of these tests will pass with a false positive if no error is thrown at all.

File: test/timesheet.js [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+            try{
+                await timesheet.validate();
+            }
+            catch(error){
+                if(error.message !== 'offHauled is not valid') throw new Error('timesheet offhauled validation failed') ;
+            }
+
+        });
+
+        it('Timesheet Validation - yardsHauled', async function() {
```
Comment (cucucachu) [inline test/timesheet.js]: Test name says yardsHauled, but the test itself is testing offHauled.

**Review by cucucachu — CHANGES_REQUESTED** [review]
Add comments to your method. 

Validate doesn't need to be asynchronous, so neither do the tests. 

Other minor changes needed in the tests.

Otherwise the validate() function is written according to the acceptance criteria, and all test cases are tested.

File: source/timesheet.js (hunk 4 above)
Comment (gma1083) [reply source/timesheet.js]: fixed.

File: test/job.js (hunk 6 above)
Comment (gma1083) [reply test/job.js]: Reverted. Done.

File: test/timesheet.js (hunk 7 above)
Comment (gma1083) [reply test/timesheet.js]: Fixed.

File: test/timesheet.js (hunk 8 above)
Comment (gma1083) [reply test/timesheet.js]: Fixed.

File: test/timesheet.js (hunk 9 above)
Comment (gma1083) [reply test/timesheet.js]: Fixed by throwing a new error ('Error expected but not thrown in Timesheet Validation - [Test]') after the timesheet.validate() call that will only be executed if no error is thrown.

File: test/timesheet.js (hunk 10 above)
Comment (gma1083) [reply test/timesheet.js]: Fixed.

File: source/timesheet.js (hunk 1 above)
Comment (gma1083) [reply source/timesheet.js:24]: Noted - but kept.

File: source/timesheet.js (hunk 2 above)
Comment (gma1083) [reply source/timesheet.js:28]: Noted - but kept.

File: source/timesheet.js (hunk 3 above)
Comment (gma1083) [reply source/timesheet.js:29]: Noted - but kept.

File: source/timesheet.js (hunk 5 above)
Comment (gma1083) [reply source/timesheet.js:32]: Added.

**Comment by cucucachu** [comment]
Looks good now. Good to merge with master if tests are passing. Code review complete.
